README.md
# Elastic Widgets
Set of Flutter widgets built using physics based animations.  
# Widgets
| Seek Bar | Range Picker |
|--|--|
| ![seekbar gif](https://github.com/Abiri99/elastic-widgets/blob/master/repo_files/gif/seekbar.gif?raw=true) | ![enter image description here](https://github.com/Abiri99/elastic-widgets/blob/master/repo_files/gif/range-picker.gif?raw=true) |
## Installation
##### Add this to your package's pubspec.yaml file
```yml
dependencies:
 elastic_widgets: 0.0.1
 ```
## Usage
 ### Seek bar:
 ``` dart
 ElasticSeekBar(
	 valueListener: (value) {
	   print("slider value: $value");
	 },
	 size: Size(300, 100),
	 stretchRange: 50.0,
	 minValue: 0,
	 maxValue: 100,
	 circleRadius: 12,
	 thinLineStrokeWidth: 3,
	 thickLineStrokeWidth: 4,
	 thickLineColor: Colors.blue,
	 thinLineColor: Colors.blueGrey,
	 bounceDuration: Duration(seconds: 1),
	 stiffness: 300,
	 dampingRatio: 5,
),
 ```
 ### Range picker:
 ``` dart
 ElasticRangePicker(
	 valueListener: (firstValue, secondValue) {
	   print("range picker first value: $firstValue");
	   print("range picker second value: $secondValue");
	 },
	 size: Size(300, 100),
	 stretchRange: 50.0,
	 minValue: 0,
	 maxValue: 100,
	 circleRadius: 12,
	 thinLineStrokeWidth: 3,
	 thickLineStrokeWidth: 4,
	 thickLineColor: Colors.orange,
	 thinLineColor: Colors.blueGrey,
	 bounceDuration: Duration(seconds: 1),
	 stiffness: 300,
	 dampingRatio: 5,
),
 ```
## Planned widgets
 - [ ] Radio button
 - [ ] Checkbox
 - [ ] Dialog
 - [ ] Button
---

**Made with :heart: for Flutter community** - **Pull requests are welcome** :sparkles:

keywords: widget, elastic, slider, seekbar, seek bar, range picker, rangepicker

CHANGELOG.md
## [0.0.3] - 2020-07-12

* Improved health
* Updated readme file

## [0.0.2] - 2020-07-12

* Improved health
* Updated readme file

## [0.0.1] - 2020-07-11

* Added Slider and RangePicker widgets.


test/elastic_widgets_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:elastic_widgets/elastic_widgets.dart';

void main() {
  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/util/utils.dart
import 'package:dartx/dartx.dart';

extension DoubleExtention on double {
  double coerceHorizontal(
    var trackStartX,
    var trackEndX,
  ) {
    return this.coerceAtMost(trackEndX).coerceAtLeast(trackStartX);
  }

  double coerceVertical(
    var trackY,
    var stretchRange,
  ) {
    return this
        .coerceAtMost(trackY + stretchRange)
        .coerceAtLeast(trackY - stretchRange);
  }

  double coerceToStretchRange(var x, var height, var width, var stretchRange,
      var trackStartX, var trackEndX) {
    if (this == 0) return this;
    if (this < 0) {
      return this.coerceAtLeast(x <= width / 2
          ? -(((2 * stretchRange) * (x - trackStartX)) /
                  (width - (2 * trackStartX))) +
              (0)
          : -(((2 * stretchRange) * (x - trackEndX)) /
                  (width - (2 * trackEndX))) +
              (0));
    } else {
      return this.coerceAtMost(x < width / 2
          ? (((2 * (stretchRange + height / 2) - height) * (x - trackStartX)) /
                  (width - (2 * trackStartX))) +
              (0)
          : (((2 * (stretchRange + height / 2) - height) * (x - trackEndX)) /
                  (width - (2 * trackEndX))) +
              (0));
    }
  }

  double coerceToStretchRangeWithPivotAtTheMiddle(
    var x,
    var stretchRange,
    var height,
    var startX,
    var endX,
  ) {
    if (this < 0) {
      return this.coerceAtLeast(x < (endX + startX) / 2
          ? -(((2 * (stretchRange + height / 2) - height) * (x - startX)) /
                  ((endX + startX) - (2 * startX))) +
              (0)
          : -(((2 * (stretchRange + height / 2) - height) * (x - endX)) /
                  ((endX + startX) - (2 * endX))) +
              (0));
    } else {
      return this.coerceAtMost(x < (endX + startX) / 2
          ? (((2 * (stretchRange + height / 2) - height) * (x - startX)) /
                  ((endX + startX) - (2 * startX))) +
              (0)
          : (((2 * (stretchRange + height / 2) - height) * (x - endX)) /
                  ((endX + startX) - (2 * endX))) +
              (0));
    }
  }
}


lib/src/seekbar/elastic_seekbar.dart
import 'package:elastic_widgets/src/seekbar/seekbar_painter.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/physics.dart';
import '../util/utils.dart';

// ignore: must_be_immutable
class ElasticSeekBar extends StatefulWidget {
  // A function passed from parent which notifies itself from the current value of seek bar
  final Function(String value) valueListener;

  // Size of seek bar
  final Size size;

  // Minimum value of seek bar
  final double minValue;

  // Maximum value of seek bar
  final double maxValue;

  // How much seek bar stretches in vertical axis
  double stretchRange;

  // Thickness of progress line and thumb
  final double thickLineStrokeWidth;

  // Thickness of default line
  final double thinLineStrokeWidth;

  // Radius of thumb
  final double circleRadius;

  // Color of progress line and thumb
  Color thickLineColor;

  // Color of default line
  Color thinLineColor;

  // Speed of bouncing animation
  final Duration bounceDuration;

  // Stiffness of slider while animating
  final double stiffness;

  // Damping ratio of slider while animating
  final double dampingRatio;

  ElasticSeekBar({
    Key key,
    @required this.valueListener,
    @required this.size,
    this.stretchRange,
    this.minValue = 0,
    this.maxValue = 100,
    this.circleRadius = 12,
    this.thickLineStrokeWidth = 4,
    this.thinLineStrokeWidth = 3,
    this.thickLineColor,
    this.thinLineColor,
    this.bounceDuration,
    this.stiffness = 300,
    this.dampingRatio = 8,
  }) : super(key: key) {
    if (thickLineColor == null) thickLineColor = Color(0xff1f3453);
    if (thinLineColor == null) thinLineColor = Color(0xff1f3453).withAlpha(80);
    if (stretchRange == null)
      stretchRange = size.height / 2 - circleRadius - thickLineStrokeWidth / 2;
  }

  @override
  _ElasticSeekBarState createState() => _ElasticSeekBarState();
}

class _ElasticSeekBarState extends State<ElasticSeekBar>
    with SingleTickerProviderStateMixin {
  double value;

  double thumbY;
  double thumbX;

  double trackStartX;
  double trackEndX;
  double trackY;

  AnimationController _controller;

  Animation<double> _animation;

  bool touched;

  @override
  void initState() {
    _controller = AnimationController(vsync: this, upperBound: 500);

    _controller.addListener(controllerListener);

    touched = false;
    value = (widget.maxValue - widget.minValue) / 2;
    thumbY = 0;
    thumbX = widget.size.width / 2;
    trackY = widget.size.height / 2;
    trackEndX = widget.size.width -
        widget.circleRadius -
        widget.thickLineStrokeWidth / 2;
    trackStartX = widget.circleRadius + widget.thickLineStrokeWidth / 2;
    super.initState();
  }

  controllerListener() {
    setState(() {
      thumbY = _animation.value;
    });
  }

  String getCurrentValue() {
    if (thumbX <= trackStartX) return widget.minValue.toString();
    if (thumbX >= trackEndX) return widget.maxValue.toString();
    return (((thumbX - trackStartX) / (trackEndX - trackStartX)) *
            (widget.maxValue - widget.minValue))
        .toString();
  }

  runAnimation(Offset pixelsPerSecond, Size size) {
    _animation = _controller.drive(Tween<double>(
      begin: thumbY,
      end: 0.0,
    ));
    var spring = SpringDescription(
      mass: 1.0,
      stiffness: widget.stiffness,
      damping: widget.dampingRatio,
    );

    final unitsPerSecondX = pixelsPerSecond.dx / size.width;
    final unitsPerSecondY = pixelsPerSecond.dy / size.height;
    final unitsPerSecond = Offset(unitsPerSecondX, unitsPerSecondY);
    final unitVelocity = unitsPerSecond.distance;

    final simulation = SpringSimulation(spring, 0, 1, -unitVelocity);

    _controller.animateWith(simulation);
  }

  int detectNode(var gestureDetectorDetails) {
    if (touched) return 0;
    if (gestureDetectorDetails.localPosition.dy >=
//            trackY -
            widget.size.height / 2 -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dy <=
//            trackY -
            widget.size.height / 2 +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dx >=
            thumbX - widget.circleRadius - widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dx <=
            thumbX + widget.circleRadius + widget.thickLineStrokeWidth / 2) {
      return 0;
    }
    return -1;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.size.height,
      width: widget.size.width,
      child: GestureDetector(
        onPanDown: (details) {
          var node = detectNode(details);
          if (node == 0) {
            setState(() {
              touched = true;
            });
            _controller.stop();
          }
        },
        onPanUpdate: (DragUpdateDetails dragUpdateDetails) {
          var node = detectNode(dragUpdateDetails);
          if (node == 0 || touched) {
            touched = true;
            RenderBox box = context.findRenderObject();
            var touchPoint =
                box.globalToLocal(dragUpdateDetails.globalPosition);
            if (dragUpdateDetails.localPosition.dx <= 0) {
              touchPoint = new Offset(0, 0.0);
            }
            if (touchPoint.dx >= context.size.width) {
              touchPoint = new Offset(context.size.width, 0);
            }
            if (touchPoint.dy <= 0) {
              touchPoint = new Offset(touchPoint.dx, 0.0);
            }
            if (touchPoint.dy >= context.size.height) {
              touchPoint = new Offset(touchPoint.dx, context.size.height);
            }
            setState(() {
              thumbX = touchPoint.dx.coerceHorizontal(trackStartX, trackEndX);
              thumbY = (touchPoint.dy - widget.size.height / 2)
                  .coerceVertical(
                      0,
                      widget.size.height / 2 -
                          widget.circleRadius -
                          widget.thickLineStrokeWidth / 2)
                  .coerceToStretchRange(
                      thumbX,
                      widget.size.height,
                      widget.size.width,
                      widget.stretchRange,
                      trackStartX,
                      trackEndX);
            });
            widget.valueListener(getCurrentValue());
          }
        },
        onPanEnd: (DragEndDetails dragEndDetails) {
          touched = false;
          runAnimation(dragEndDetails.velocity.pixelsPerSecond, widget.size);
        },
        child: CustomPaint(
          size: Size(widget.size.width, widget.size.height),
          painter: SeekBarPainter(
            thumbX: thumbX,
            thumbY: thumbY,
            width: widget.size.width,
            height: widget.size.height,
            touched: touched,
            thickLineColor: widget.thickLineColor,
            thickLineStrokeWidth: widget.thickLineStrokeWidth,
            thinLineColor: widget.thinLineColor,
            thinLineStrokeWidth: widget.thinLineStrokeWidth,
            circleRadius: widget.circleRadius,
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    _controller.removeListener(controllerListener);
    _controller.dispose();
    super.dispose();
  }
}


lib/src/seekbar/seekbar_painter.dart
import 'package:flutter/material.dart';

class SeekBarPainter extends CustomPainter {
  double x1;
  double y1;
  double x2;
  double y2;

  double thumbX;
  double thumbY;

  Path path;

  bool touched;

  double width;
  double height;

  double thickLineStrokeWidth;
  double thinLineStrokeWidth;

  Color thickLineColor;
  Color thinLineColor;

  double circleRadius;

  Paint painter;

  SeekBarPainter({
    @required this.thumbX,
    @required this.thumbY,
    @required this.width,
    @required this.height,
    @required this.thickLineStrokeWidth,
    @required this.thinLineStrokeWidth,
    @required this.thickLineColor,
    @required this.thinLineColor,
    @required this.circleRadius,
    this.touched = false,
  }) {
    path = Path();
    path.reset();
    painter = Paint();
  }

  double get trackEndX {
    return width - circleRadius - thickLineStrokeWidth / 2;
  }

  double get trackStartX {
    return circleRadius + thickLineStrokeWidth / 2;
  }

  double get trackY {
    return height / 2;
  }

  @override
  void paint(Canvas canvas, Size size) {
    var mThumbY = thumbY + height / 2;

    // Progress line
    path.reset();
    painter = Paint()
      ..color = thickLineColor
      ..style = PaintingStyle.stroke
      ..strokeCap = StrokeCap.round
      ..strokeWidth = thickLineStrokeWidth;

    x1 = (thumbX + trackStartX) / 2;
    y1 = height / 2;
    x2 = x1;
    y2 = mThumbY;

    path.moveTo(trackStartX, height / 2);
    path.cubicTo(x1, y1, x2, y2, thumbX - circleRadius, mThumbY);
    if (thumbX - circleRadius >= trackStartX) canvas.drawPath(path, painter);

    // Default line
    painter = Paint()
      ..color = thinLineColor
      ..style = PaintingStyle.stroke
      ..strokeCap = StrokeCap.round
      ..strokeWidth = thinLineStrokeWidth;

    path.reset();
    path.moveTo(thumbX + circleRadius, mThumbY);

    x1 = (thumbX + trackEndX) / 2;
    y1 = mThumbY;
    x2 = x1;
    y2 = height / 2;

    path.cubicTo(x1, y1, x2, y2, trackEndX, trackY);

    if (thumbX + circleRadius <= trackEndX) canvas.drawPath(path, painter);

    // Circle inside
    painter = Paint()
      ..color = thickLineColor.withOpacity(touched ? 0.9 : 0.4)
      ..style = PaintingStyle.fill;

    path.reset();
    canvas.drawCircle(Offset(thumbX, mThumbY),
        circleRadius - thickLineStrokeWidth / 2, painter);

    // Circle border
    painter = Paint()
      ..color = thickLineColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = thickLineStrokeWidth;

    canvas.drawCircle(Offset(thumbX, mThumbY), circleRadius, painter);

    path.reset();
  }

  @override
  bool shouldRepaint(CustomPainter oldPainter) {
    return oldPainter != this;
  }
}


lib/src/range_picker/elastic_range_picker.dart
import 'package:elastic_widgets/src/range_picker/range_picker_painter.dart';
import 'package:flutter/material.dart';
import 'package:flutter/physics.dart';
import '../util/utils.dart';

// ignore: must_be_immutable
class ElasticRangePicker extends StatefulWidget {
  // Size of widget
  final Size size;

  // A function passed from parent which notifies itself from the current value of range picker
  final Function(double firstVal, double secVal) valueListener;
  final double minValue;
  final double maxValue;

  // How much seek bar stretches in vertical axis
  double stretchRange;

  // Thickness of progress line and thumb
  final double thickLineStrokeWidth;

  // Thickness of default line
  final double thinLineStrokeWidth;

  // Radius of thumb
  final double circleRadius;

  // Color of progress line and thumb
  Color thickLineColor;

  // Color of default line
  Color thinLineColor;

  // Speed of bouncing animation
  final Duration bounceDuration;

  // Stiffness of slider while animating
  final double stiffness;

  // Damping ratio of slider while animating
  final double dampingRatio;

  ElasticRangePicker({
    this.size,
    this.valueListener,
    this.stretchRange,
    this.minValue = 0,
    this.maxValue = 100,
    this.circleRadius = 12,
    this.thickLineStrokeWidth = 4,
    this.thinLineStrokeWidth = 3,
    this.thickLineColor,
    this.thinLineColor,
    this.bounceDuration,
    this.stiffness = 300,
    this.dampingRatio = 8,
  }) {
    if (thickLineColor == null) thickLineColor = Color(0xff1f3453);
    if (thinLineColor == null) thinLineColor = Colors.blueGrey;
    if (stretchRange == null)
      stretchRange = size.height / 2 - circleRadius - thickLineStrokeWidth / 2;
  }

  @override
  _ElasticRangePickerState createState() => _ElasticRangePickerState();
}

class _ElasticRangePickerState extends State<ElasticRangePicker>
    with TickerProviderStateMixin {
  double firstThumbY;
  double firstThumbX;
  double secThumbY;
  double secThumbX;

  double trackStartX;
  double trackEndX;
  double trackY;

  double firstValue;
  double secondValue;

  bool firstNodeTouched;
  bool secNodeTouched;

  AnimationController _firstController;
  AnimationController _secondController;

  Animation<double> _firstAnimation;
  Animation<double> _secondAnimation;

  @override
  void initState() {
    _firstController = AnimationController(vsync: this, upperBound: 500);
    _secondController = AnimationController(vsync: this, upperBound: 500);

    _firstController.addListener(firstControllerListener);
    _secondController.addListener(secondControllerListener);

    firstNodeTouched = false;
    secNodeTouched = false;
    firstValue = (widget.maxValue - widget.minValue) / 3;
    secondValue = 2 * (widget.maxValue - widget.minValue) / 3;
    firstThumbY = 0;
    secThumbY = 0;
    firstThumbX = widget.size.width / 3;
    secThumbX = 2 * widget.size.width / 3;
    trackEndX = widget.size.width -
        widget.circleRadius -
        widget.thickLineStrokeWidth / 2;
    trackStartX = widget.circleRadius + widget.thickLineStrokeWidth / 2;
    super.initState();
  }

  firstControllerListener() {
    setState(() {
      firstThumbY = _firstAnimation.value;
    });
  }

  secondControllerListener() {
    setState(() {
      secThumbY = _secondAnimation.value;
    });
  }

  double getFirstValue() {
    if (firstThumbX <= trackStartX) return widget.minValue;
    if (firstThumbX >= trackEndX) return widget.maxValue;
    return (((firstThumbX - trackStartX) / (trackEndX - trackStartX)) *
        (widget.maxValue - widget.minValue));
  }

  double getSecValue() {
    if (secThumbX <= trackStartX) return widget.minValue;
    if (secThumbX >= trackEndX) return widget.maxValue;
    return (((secThumbX - trackStartX) / (trackEndX - trackStartX)) *
        (widget.maxValue - widget.minValue));
  }

  runFirstThumbAnimation(Offset pixelsPerSecond, Size size) {
    _firstAnimation = _firstController.drive(Tween<double>(
      begin: firstThumbY,
      end: 0.0,
    ));
    var spring = SpringDescription(
      mass: 1.0,
      stiffness: widget.stiffness,
      damping: widget.dampingRatio,
    );

    final unitsPerSecondX = pixelsPerSecond.dx / size.width;
    final unitsPerSecondY = pixelsPerSecond.dy / size.height;
    final unitsPerSecond = Offset(unitsPerSecondX, unitsPerSecondY);
    final unitVelocity = unitsPerSecond.distance;

    final simulation = SpringSimulation(spring, 0, 1, -unitVelocity);

    _firstController.animateWith(simulation);
  }

  runSecondThumbAnimation(Offset pixelsPerSecond, Size size) {
    _secondAnimation = _secondController.drive(Tween<double>(
      begin: secThumbY,
      end: 0.0,
    ));
    var spring = SpringDescription(
      mass: 1.0,
      stiffness: widget.stiffness,
      damping: widget.dampingRatio,
    );

    final unitsPerSecondX = pixelsPerSecond.dx / size.width;
    final unitsPerSecondY = pixelsPerSecond.dy / size.height;
    final unitsPerSecond = Offset(unitsPerSecondX, unitsPerSecondY);
    final unitVelocity = unitsPerSecond.distance;

    final simulation = SpringSimulation(spring, 0, 1, -unitVelocity);

    _secondController.animateWith(simulation);
  }

  int detectNodeByTouchpoint(var touchpoint) {
    if (firstNodeTouched) return 0;
    if (secNodeTouched) return 1;
    if (touchpoint.dy >=
            widget.size.height / 2 -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        touchpoint.dy <=
            widget.size.height / 2 +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2 &&
        touchpoint.dx >=
            firstThumbX -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        touchpoint.dx <=
            firstThumbX +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2) {
      return 0;
    } else if (touchpoint.dy >=
//            trackY -
            widget.size.height / 2 -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        touchpoint.dy <=
//            trackY -
            widget.size.height / 2 +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2 &&
        touchpoint.dx >=
            secThumbX - widget.circleRadius - widget.thickLineStrokeWidth / 2 &&
        touchpoint.dx <=
            secThumbX + widget.circleRadius + widget.thickLineStrokeWidth / 2) {
      return 1;
    }
    return -1;
  }

  int detectNode(var gestureDetectorDetails) {
    if (firstNodeTouched) return 0;
    if (secNodeTouched) return 1;
    if (gestureDetectorDetails.localPosition.dy >=
//            trackY -
            widget.size.height / 2 -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dy <=
//            trackY -
            widget.size.height / 2 +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dx >=
            firstThumbX -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dx <=
            firstThumbX +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2) {
      return 0;
    } else if (gestureDetectorDetails.localPosition.dy >=
//            trackY -
            widget.size.height / 2 -
                widget.circleRadius -
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dy <=
//            trackY -
            widget.size.height / 2 +
                widget.circleRadius +
                widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dx >=
            secThumbX - widget.circleRadius - widget.thickLineStrokeWidth / 2 &&
        gestureDetectorDetails.localPosition.dx <=
            secThumbX + widget.circleRadius + widget.thickLineStrokeWidth / 2) {
      return 1;
    }
    return -1;
  }

  @override
  Widget build(BuildContext context) {
    var size = MediaQuery.of(context).size;
    return Container(
      height: widget.size.height,
      width: widget.size.width,
      child: GestureDetector(
        onPanDown: (details) {
          var node = detectNode(details);
          if (node == 0) {
            setState(() {
              firstNodeTouched = true;
            });
            _firstController.stop();
          }
          if (node == 1) {
            setState(() {
              secNodeTouched = true;
            });
            _secondController.stop();
          }
        },
        onPanUpdate: (DragUpdateDetails dragUpdateDetails) {
          RenderBox box = context.findRenderObject();
          var touchPoint = box.globalToLocal(dragUpdateDetails.globalPosition);
          if (touchPoint.dx <= 0) {
            touchPoint = new Offset(0, 0.0);
          }
          if (touchPoint.dx >= trackEndX) {
            touchPoint = new Offset(trackEndX, 0);
          }
          if (touchPoint.dy <= 0) {
            touchPoint = new Offset(touchPoint.dx, 0.0);
          }
          if (touchPoint.dy >= context.size.height) {
            touchPoint = new Offset(touchPoint.dx, context.size.height);
          }
          var node = detectNodeByTouchpoint(touchPoint);
          if (node == 0) {
            firstNodeTouched = true;
            setState(
              () {
                firstThumbX = touchPoint.dx.coerceHorizontal(
                  trackStartX,
                  trackEndX -
                      2 *
                          (widget.circleRadius +
                              widget.thickLineStrokeWidth / 2),
                );
                if (firstThumbX >=
                    secThumbX -
                        2 *
                            (widget.circleRadius +
                                widget.thickLineStrokeWidth / 2)) {
                  secThumbX = (firstThumbX +
                          2 *
                              (widget.circleRadius +
                                  widget.thickLineStrokeWidth / 2))
                      .coerceHorizontal(
                    trackStartX +
                        2 *
                            (widget.circleRadius +
                                widget.thickLineStrokeWidth / 2),
                    trackEndX,
                  );
                }
                firstThumbY = (touchPoint.dy - widget.size.height / 2)
                    .coerceVertical(
                      0,
                      widget.size.height / 2 -
                          widget.circleRadius -
                          widget.thickLineStrokeWidth / 2,
                    )
                    .coerceToStretchRange(
                      firstThumbX,
                      widget.size.height,
                      secThumbX,
                      widget.stretchRange,
                      trackStartX,
                      secThumbX -
                          widget.circleRadius -
                          widget.thickLineStrokeWidth / 2,
                    );
              },
            );
            widget.valueListener(
              getFirstValue(),
              getSecValue(),
            );
          } else if (node == 1) {
            secNodeTouched = true;
            setState(
              () {
                secThumbX = touchPoint.dx.coerceHorizontal(
                  trackStartX +
                      2 *
                          (widget.circleRadius +
                              widget.thickLineStrokeWidth / 2),
                  trackEndX,
                );
                if (secThumbX <=
                    firstThumbX +
                        2 *
                            (widget.circleRadius +
                                widget.thickLineStrokeWidth / 2)) {
                  firstThumbX = (secThumbX -
                          2 *
                              (widget.circleRadius +
                                  widget.thickLineStrokeWidth / 2))
                      .coerceHorizontal(
                    trackStartX,
                    trackEndX -
                        2 *
                            (widget.circleRadius +
                                widget.thickLineStrokeWidth / 2),
                  );
                }
                secThumbY = (touchPoint.dy - widget.size.height / 2)
                    .coerceVertical(
                        0,
                        widget.size.height / 2 -
                            widget.circleRadius -
                            widget.thickLineStrokeWidth / 2)
                    .coerceToStretchRangeWithPivotAtTheMiddle(
                      secThumbX,
                      widget.stretchRange,
                      widget.size.height,
                      firstThumbX,
                      trackEndX,
                    );
              },
            );
            widget.valueListener(getFirstValue(), getSecValue());
          }
        },
        onPanEnd: (DragEndDetails dragEndDetails) {
          runFirstThumbAnimation(dragEndDetails.velocity.pixelsPerSecond, size);
          runSecondThumbAnimation(
              dragEndDetails.velocity.pixelsPerSecond, size);
          setState(
            () {
              firstNodeTouched = false;
              secNodeTouched = false;
            },
          );
        },
        child: CustomPaint(
          size: Size(
            widget.size.width,
            widget.size.height,
          ),
          painter: RangePickerPainter(
            firstThumbX: firstThumbX,
            firstThumbY: firstThumbY,
            secThumbX: secThumbX,
            secThumbY: secThumbY,
            circleRadius: widget.circleRadius,
            thickLineStrokeWidth: widget.thickLineStrokeWidth,
            thinLineStrokeWidth: widget.thinLineStrokeWidth,
            thickLineColor: widget.thickLineColor,
            thinLineColor: widget.thinLineColor,
            width: widget.size.width,
            height: widget.size.height,
            firstNodeTouched: firstNodeTouched,
            secNodeTouched: secNodeTouched,
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:elastic_widgets/elastic_widgets.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ElasticSeekBar(
              valueListener: (value) {
                print("slider value: $value");
              },
              size: Size(300, 100),
              stretchRange: 50.0,
              minValue: 0,
              maxValue: 100,
              circleRadius: 12,
              thinLineStrokeWidth: 3,
              thickLineStrokeWidth: 4,
              thickLineColor: Colors.blue,
              thinLineColor: Colors.blueGrey,
              bounceDuration: Duration(seconds: 1),
              stiffness: 300,
              dampingRatio: 5,
            ),
            SizedBox(
              height: 64,
            ),
            ElasticRangePicker(
              valueListener: (firstValue, secondValue) {
                print("range picker first value: $firstValue");
                print("range picker second value: $secondValue");
              },
              size: Size(300, 100),
              stretchRange: 50.0,
              minValue: 0,
              maxValue: 100,
              circleRadius: 12,
              thinLineStrokeWidth: 3,
              thickLineStrokeWidth: 4,
              thickLineColor: Colors.orange,
              thinLineColor: Colors.blueGrey,
              bounceDuration: Duration(seconds: 1),
              stiffness: 300,
              dampingRatio: 5,
            ),
          ],
        ),
      ),// This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
